File: Data Structures & Algorithms/valid-tree/submission-3.py
class Solution:
    def validTree(self, n: int, edges: List[List[int]]) -> bool:
        
        #making the adj list - undirected graph
        adjList = {}
        for i in range(n):
            adjList[i] = []
        for edge in edges:
            for i in range(len(edge)):
                if i == 0:
                    continue
                #add the forward edge
                adjList[edge[0]].append(edge[i])
                #add the backward edge
                adjList[edge[i]].append(edge[0])

        seen = set()
        def BFS(node):
            queue = []
            queue.append(node)

            while queue:
                v = queue.pop(0)
                
                for neighbors in adjList[v]:
                    if neighbors not in seen:
                        queue.append(neighbors)
                    
                seen.add(v)
        
        #if connected, all nodes must be in seen
        BFS(0)

        print(seen)
        
        #check if all nodes are connected or not
        for i in range(n):
            if i not in seen:
                return False

        #simple check. If there are more than n-1 edges, there is a cycle and return False. Else return True
        count = 0
        for i in range(n):
            count = count + len(adjList[i])
        
        if count > (2*(n-1)):
            return False
        else:
            return True

File: Data Structures & Algorithms/lemonade-change/submission-3.cpp
class Solution {
public:
    bool lemonadeChange(vector<int>& bills) {
        vector<int> count = {0,0,0};

        for (int& b: bills) {
            if (b == 5) { count[0]++; }
            if (b == 10) {
                if (count[0] <  1) { return false; }
                count[0]--;
                count[1]++;
            }
            if (b == 20) {
                if (count[0] >= 1 && count[1] >= 1) {
                    count[0]--; count[1]--;
                    count[2]++;
                } else if (count[0] >= 3) {
                    count[0] -= 3;
                    count[2]++;
                } else return false;
            }
        }
        return true;
    }
};